modules/ml/src/main/java/org/apache/ignite/ml/trees/trainers/columnbased/ColumnDecisionTreeTrainer.java

```diff
@@ -0,0 +1,559 @@
+            IgniteCache<RegionKey, List<RegionProjection>> projsCache = ProjectionsCache.getOrCreate(Ignition.localIgnite());
+
+            return () -> IntStream.range(0, (rc - 1) / BLOCK_SIZE + 1).
+                mapToObj(rBIdx -> ProjectionsCache.key(0, rBIdx, input.affinityKey(0, Ignition.localIgnite()), uuid)).
+                map(k -> (Cache.Entry<RegionKey, List<RegionProjection>>)new CacheEntryImpl<>(k, projsCache.localPeek(k))).iterator();
+        };
+
+        Map<Integer, Double> vals = CacheUtils.reduce(prjsCache.getName(), ignite,
+            (TrainingContext ctx, Cache.Entry<RegionKey, List<RegionProjection>> e, Map<Integer, Double> m) -> {
+                int regBlockIdx = e.getKey().regionBlockIndex();
+
+                if (e.getValue() != null) {
+                    for (int i = 0; i < e.getValue().size(); i++) {
+                        int regIdx = regBlockIdx * BLOCK_SIZE + i;
+                        RegionProjection reg = e.getValue().get(i);
+
+                        Double res = regCalc.apply(Arrays.stream(reg.sampleIndexes()).mapToDouble(s -> ctx.labels()[s]));
+                        m.put(regIdx, res);
+                    }
+                }
+
+                return m;
+            },
+            () -> ContextCache.getOrCreate(Ignition.localIgnite()).get(uuid),
+            featZeroRegs,
+            (infos, infos2) -> {
+                Map<Integer, Double> res = new HashMap<>();
+                res.putAll(infos);
+                res.putAll(infos2);
+                return res;
+            },
+            HashMap::new
+        );
+
+        int i = 0;
+        for (TreeTip tip : tips) {
+            tip.leafSetter.accept(new Leaf(vals.get(i)));
+            i++;
+        }
+
+        ProjectionsCache.clear(featuresCnt, rc, input::affinityKey, uuid, ignite);
+        ContextCache.getOrCreate(ignite).remove(uuid);
+        FeaturesCache.clear(featuresCnt, input::affinityKey, uuid, ignite);
+        SplitCache.clear(featuresCnt, input::affinityKey, uuid, ignite);
+
+        return new DecisionTreeModel(root.s);
+    }
+
+    /**
+     * Find the best split in the form (feature index, (index of region with the best split, impurity of region with the
+     * best split)).
+     *
+     * @param featuresCnt Count of features.
+     * @param affinity Affinity function.
+     * @param trainingUUID UUID of training.
+     * @return Best split in the form (feature index, (index of region with the best split, impurity of region with the
+     * best split)).
+     */
+    private IgniteBiTuple<Integer, IgniteBiTuple<Integer, Double>> findBestSplitIndexForFeatures(int featuresCnt,
+        IgniteBiFunction<Integer, Ignite, Object> affinity,
+        UUID trainingUUID) {
+        Set<Integer> featureIndexes = IntStream.range(0, featuresCnt).boxed().collect(Collectors.toSet());
+
+        return CacheUtils.reduce(COLUMN_DECISION_TREE_TRAINER_SPLIT_CACHE_NAME, ignite,
```

Comment:
It's unreadable piece of lambdas. Extract to variables. give them clear names or comment it